### PR TITLE
Cancel crafting druing component selection

### PR DESCRIPTION
**Pull request by JornVernee**
The idea is: when crafting (especially when repeating a craft), sometimes you expect a component to be there but it isn't (because it got used up). There are still other components that can be used, but you might not want to. So I added the ability to cancel crafting when this happens.

Example:
I'm crafting some item that can ether use a welder or some other components. When I repeat the craft (using '-'), I can not select my welder, because it ran out of batteries, and I don't want to waste the other components. Now I can cancel crafting at this point.

Right now I only implemented it for regular crafting. It might be useful to add it to construction and vehicle crafting too. Or to add a warning when repeating a craft, that the component types you previously used are no longer available. I'm fine with making that too, but I first wanted to get some feedback on the idea.

Note: I'm fairly new to forking/merging/etc. so the commits are a bit messy, sorry.

### CODE REVIEW

File: src/crafting.cpp [hunk 1]
```diff
@@ -1603,12 +1648,12 @@ void player::complete_craft()
     if (making->difficulty != 0 && diff_roll > skill_roll * (1 + 0.1 * rng(1, 5))) {
         add_msg(m_bad, _("You fail to make the %s, and waste some materials."),
                 item::nname(making->result).c_str());
-        for (const auto &it : making->requirements.components) {
-            consume_items(it, batch_size);
+		for (int i = 0; i < making->requirements.components.size(); i++) {
```
Comment (Coolthulhu) [inline src/crafting.cpp]: Don't use tabs, use 4 spaces instead.

File: src/crafting.cpp [hunk 2]
```diff
@@ -1746,219 +1791,275 @@ void set_item_inventory(item &newit)
...
+		else if (map_has.size() == 1) {
+			selected.use_from = usage::use_from_map;
+			selected.comp = map_has[0];
+		}
+		else {
+			selected.use_from = usage::use_from_both;
+			selected.comp = mixed[0];
+		}
+	}
+	else { // Let the player pick which component they want to use
```
Comment (Coolthulhu) [inline src/crafting.cpp]: Keep `} else {` on one line

File: src/cursesdef.h [hunk 3]
```diff
@@ -1,7 +1,7 @@
 #ifndef CURSESDEF_H
 #define CURSESDEF_H
 
-#if (defined TILES || defined _WIN32 || defined WINDOWS)
+#if (defined TILES || defined SDLTILES || defined _WIN32 || defined WINDOWS)
```
Comment (Coolthulhu) [inline src/cursesdef.h]: Why is this change in this PR?
I'm pretty sure SDLTILES was obsoleted and thus shouldn't be referenced anywhere.

**Comment by JornVernee** [comment]
looks like '-Werror' is not in the .cbp? I'm fixing now.

File: src/cursesdef.h (hunk 3 above)
Comment (JornVernee) [inline src/cursesdef.h]: ah yes, something went wrong so I manually tried to revert it, must have missed this one

File: src/crafting.cpp (hunk 2 above)
Comment (JornVernee) [inline src/crafting.cpp]: fine by me, this code was already there though

File: src/crafting.cpp [hunk 4]
```diff
@@ -1625,12 +1670,12 @@ void player::complete_craft()
...
     std::list<item> used;
-    for (const auto &it : making->requirements.components) {
-        std::list<item> tmp = consume_items(it, batch_size);
+    for (int i = 0; i < making->requirements.components.size(); i++) {
+        std::list<item> tmp = consume_items(comp_selections[i], batch_size); // again, we consume the items we selected earlier
         used.splice(used.end(), tmp);
     }
-    for (const auto &it : making->requirements.tools) {
-        consume_tools(it, batch_size);
+	for (int i = 0; i < making->requirements.tools.size(); i++) {
```
Comment (Coolthulhu) [inline src/crafting.cpp]: Any reason you're changing all those `for` loops into indexed ones?
Couldn't you iterate over `tool_selections` here? That would be more readable than having max index depend on `making->requirements.tools.size()`, but the actual indexed variable be from `tool_selections`.

Comment (JornVernee) [inline src/crafting.cpp]: good point, I first had the idea of some mapping of the selected component to the component group, which didn't work to well, but it flowed into this form.

**Comment by Zireael07** [comment]
Great!

File: src/crafting.cpp [hunk 5]
```diff
@@ -1746,219 +1791,277 @@ void set_item_inventory(item &newit)
     }
 }
 
+/* selection of component if a recipe requirement has multiple options (e.g. 'duct tap' or 'welder') */
+comp_selection player::select_component(const std::vector<item_comp> &components, int batch, bool can_cancel) {
+	std::vector<item_comp> player_has;
+	std::vector<item_comp> map_has;
+	std::vector<item_comp> mixed;
+
+	comp_selection selected = comp_selection();
```
Comment (BevapDin) [inline src/crafting.cpp]: Why are you calling the default constructor explicitly? Is a simple `comp_selection selected;` not good enough?

File: src/crafting.cpp [hunk 6]
```diff
@@ -1565,6 +1587,29 @@ void player::complete_craft()
...
+			activity.type = ACT_NULL;
+			return;
+		}
+		comp_selections.push_back(cs);
+	}
+
+	std::vector<tool_selection> tool_selections = std::vector<tool_selection>();
+	for (auto &it : making->requirements.tools) {
+		tool_selection ts;
+		if ((ts = select_tool(it, batch_size, DEFAULT_HOTKEYS, true)).use_from == usage::cancel) {
```
Comment (BevapDin) [inline src/crafting.cpp]: Assignment and comparison (of a member) in an if-condition is kind of unusual and not readable.

You also don't gain anything from it. `ts` has been initialized anyway. Why not split this into one line with the definition and assignment of `ts` and another line for the if statement?

File: src/crafting.cpp [hunk 7]
```diff
@@ -1565,6 +1587,29 @@ void player::complete_craft()
         skill_dice -= paws_rank_penalty * 4;
     }
 
+	/* we select the components we want to use */
+	std::vector<comp_selection> comp_selections = std::vector<comp_selection>();
```
Comment (BevapDin) [inline src/crafting.cpp]: Same here, why the explicit call to the constructor?

File: src/crafting.cpp [hunk 8]
```diff
@@ -24,6 +24,28 @@
...
+	item_comp comp;
+
+	comp_selection() : use_from(usage::use_from_none), comp(item_comp("", 0)) { }
+};
+
+struct tool_selection {
+	usage use_from;
+	tool_comp comp;
+
+	tool_selection() : use_from(usage::use_from_none), comp(tool_comp("", 0)) { }
```
Comment (BevapDin) [inline src/crafting.cpp]: The constructor could be removed and the members could be initialized in their declaration instead (if needed at all):

```C++
struct foo {
    usage use_from = use_from_none; // Btw. the usage:: prefix is not needed.
    tool_comp comp; // tool_comp actually already has a decent default constructor.
};
```

File: src/crafting.cpp [hunk 9]
```diff
@@ -24,6 +24,28 @@
...
 
+enum usage {
+	use_from_map = 1,
+	use_from_player = 2,
+	use_from_both = 1 | 2,
+	use_from_none = 4,
+	cancel = 5
+};
+
+struct comp_selection {
```
Comment (BevapDin) [inline src/crafting.cpp]: This looks as if it's the same as the `tool_selection` only with different `comp` type. It could be made into a templated class:

```C++
template<typename CompType>
struct comp_selection {
    usage use_from;
    CompType comp;
};
using item_selection = comp_selection<item_comp>;
using tool_selection = comp_selection<tool_comp>;
```

File: src/crafting.cpp [hunk 10]
```diff
@@ -1746,219 +1791,277 @@ void set_item_inventory(item &newit)
...
 
+/* selection of component if a recipe requirement has multiple options (e.g. 'duct tap' or 'welder') */
+comp_selection player::select_component(const std::vector<item_comp> &components, int batch, bool can_cancel) {
+	std::vector<item_comp> player_has;
+	std::vector<item_comp> map_has;
+	std::vector<item_comp> mixed;
+
+	comp_selection selected = comp_selection();
+	inventory map_inv;
+	map_inv.form_from_map(pos3(), PICKUP_RANGE);
```
Comment (BevapDin) [inline src/crafting.cpp]: For each list of components you create a new inventory containing all the items around your character. Have you tried this around a ***-load of items? That's why there is `player::crafting_inventory`. You may need to add an overload of that function for const player objects.

File: src/crafting.cpp (hunk 5 above)
Comment (JornVernee) [inline src/crafting.cpp]: Well wouldn't ```comp_selection selected;``` leave the values of the struct uninitialized? Isn't it safer to initialize them?

File: src/crafting.cpp (hunk 6 above)
Comment (JornVernee) [inline src/crafting.cpp]: good point, I'll add it

File: src/crafting.cpp (hunk 8 above)
Comment (JornVernee) [inline src/crafting.cpp]: yes, I wanted to do this, but I got some error like 'cannot assign to non-static member bla bla' it seems to work now though.

File: src/crafting.cpp (hunk 9 above)
Comment (JornVernee) [inline src/crafting.cpp]: I thought about doing this, but was unsure about how to add type constraints in c++.

File: src/crafting.cpp (hunk 10 above)
Comment (JornVernee) [inline src/crafting.cpp]: I think I understand what you mean, but it's really not my territory. When I started making changes this creation of ```map_inv``` was already being done for every component. It just looks weird in the diff because I encapsulated the selecting of components in a separate function.

File: src/crafting.cpp (hunk 9 above)
Comment (JornVernee) [inline src/crafting.cpp]: ```using``` ? don't you mean typedef?

File: src/crafting.cpp (hunk 5 above)
Comment (BevapDin) [inline src/crafting.cpp]: > Isn't it safer to initialize them?

That's why you have a constructor. This code:
```C++
comp_selection selected;
```
*automatically* calls the default constructor. You can see this all around the code, even two lines above, you see `std::vector<item_comp> mixed;`, which no explicit initialization. That seems to be fine, why does it bother you here?

File: src/crafting.cpp (hunk 9 above)
Comment (BevapDin) [inline src/crafting.cpp]: For what would you need constrains here? Usually, the compiler complains if the type for whatever reason doesn't work (doesn't fit the constrains):

```C++
// note: CompType needs a foo() function.
template<typename CompType>
struct comp_selection {
    usage use_from;
    CompType comp;

    void bar() {
        comp.foo(); // compiler complains here if the specific CompType doesn't have foo.
    }
};
```

That's not really optimal, but it seems good enough with the (very limited) use here.

File: src/crafting.cpp (hunk 10 above)
Comment (BevapDin) [inline src/crafting.cpp]: Ah, I see, it was already that way. In that case, interpret my words as further (maybe performance enhancing) suggesting and complete optional.

File: src/crafting.cpp (hunk 9 above)
Comment (JornVernee) [inline src/crafting.cpp]: You're right. I just got side tracked yesterday wondering how to do type constraints, not finding a quick answer, and putting off the template till later.

File: src/crafting.cpp (hunk 5 above)
Comment (JornVernee) [inline src/crafting.cpp]: That's good to know, I wasn't sure to be honest. iirc when doing ANSI C a declaration like that doesn't call a constructor.

File: src/crafting.cpp (hunk 10 above)
Comment (JornVernee) [inline src/crafting.cpp]: It seems like a good idea to put that in a separate pull request.

**Comment by Rivet-the-Zombie** [comment]
Nice!

File: src/crafting.cpp [hunk 11]
```diff
@@ -1914,51 +1968,85 @@ void player::consume_tools(const std::vector<tool_comp> &tools, int batch, const
...
+            return selected;    // and the fire goes out.
         }
 
         // Get selection via a popup menu
-        size_t selection = menu_vec(false, _("Use which tool?"), options, hotkeys) - 1;
-        if (selection < map_has.size()) {
-            long quantity = map_has[selection].count * batch;
-            g->m.use_charges(pos3(), PICKUP_RANGE, map_has[selection].type, quantity );
+        int selection = menu_vec(can_cancel, _("Use which tool?"), options, hotkeys) - 1;
+        if (selection == -1) {
```
Comment (BevapDin) [inline src/crafting.cpp]: Are you sure `menu_vec` returns 0 when canceled? The comment in output.h indicates otherwise:
```
// for the next two functions, if cancelable is true, esc returns the last option
```
If that comment is wrong, it should be corrected.

Also: I think there is a return or break messing after assigning `selected.use_from`. Currently it's always overridden with `use_from_map` or `use_from_player` a few lines below.

File: src/crafting.cpp [hunk 12]
```diff
@@ -1746,85 +1785,76 @@ void set_item_inventory(item &newit)
...
-        for( auto &map_ha : map_has ) {
-            std::string tmpStr = item::nname( map_ha.type ) + _( " (nearby)" );
+                                          // Populate options with the names of the items
+        for (auto &map_ha : map_has) {
+            std::string tmpStr = item::nname(map_ha.type) + _(" (nearby)");
             options.push_back(tmpStr);
         }
-        for( auto &player_ha : player_has ) {
-            options.push_back( item::nname( player_ha.type ) );
+        for (auto &player_ha : player_has) {
```
Comment (BevapDin) [inline src/crafting.cpp]: There are many changes to the spaces like those here, which do not respect the official style, see doc/CODE_STYLE.txt

File: src/crafting.cpp (hunk 11 above)
Comment (JornVernee) [inline src/crafting.cpp]: I has some trouble figuring out what ```menu_vec``` was returning, so I just did a printout of the value using ```debugmsg``` which returned -1 when I pressed escape. I just used that to check if we cancelled.

Alos: yes! I missed the return statement.

File: src/crafting.cpp (hunk 12 above)
Comment (JornVernee) [inline src/crafting.cpp]: got it, I didn't know there was a preference (my fault). I'll keep an eye out for it.

Comment (JornVernee) [inline src/crafting.cpp]: I must add, there's a lot of instances in the original code that also don't respect the official style.

Comment (BevapDin) [inline src/crafting.cpp]: It was established after most of the code had already been written. The existing code wasn't touched at that point because it would make too many problems. The current rule is to formatted new code and format code that is change anyway.

**Comment by kevingranade** [comment]
I like the feature you're adding, but your changes in support of that are absolutely dwarfed by a large number of changes that seem to have nothing to do with the feature, and have no rationale provided for why you're making them.  Please pick out just the changes for the feature, and if there are separate changes, they can and should go in their own PR.

**Comment by JornVernee** [comment]
Ok, I'll leave it as is then. All unrelated changes were removed with the last commit.

**Comment by JornVernee** [comment]
You're right, I got a little carried away. Most of the unrelated changes were removed with the last commit. The only one that is still there is the renaming of 'recipes' to 'recipes_by_catagory', I can change that back too but I think this renaming is a good idea, it confused me a lot when first looking at the code and seeing 'recipes' everywhere and than something like:
```    
for( auto &recipe : recipes ) {
    for( auto &elem : recipe.second ) {
            ...
    }
}
```
does a recipe contain a list of recipes within it?

That aside, I don't actually change that much, though the diff might make it seem like that. I had some trouble setting op the pull request and somehow an older version of the project got mixed in. I couldn't figure out how to 'revert to upstream' so I ended up doing it manually.

Let me try to explain what I'm actually doing a little better.

In the original code the function ```complete_craft``` did most of the work it looked something like this:
```
complete_craft {
    <gain experience>
    <consume items>
}
```
in the 'consume items' part the function the function ```player::consume_items``` was called which looked something like this:
```
consume_items {
   <select items to be consumed>
   <consume the selected items>
}
```
This was a problem, to implement cancelling during component selection, all of the component selection has to be done first. Otherwise half of the components might be consumed, and then suddenly crafting is cancelled, we would have to somehow un-consume the consumed items. So I split up ```consume_items``` into a function that selects the items and into one that consumes them, while maintaining the original function signature as well:
```
select_components {
    <select components to use>
    <return a selection object>
}

consume_items( pre-selected component ) {
     <consume the pre-selected component>
}

consume_items { // The original function signature
    <call select_components>
    <pass selection to the above consume_items>
}
```
I made this change for both the tools and items. Then I changed ```complete_craft```:
```
complete_craft {
    <select items to use>
    <gain experience>
    <consume items>
}
```
The selection should be done before anything significant, so we don't have to undo those things if we cancel.

**Comment by JornVernee** [comment]
I think it's a good idea to set up a new PR with just the changes for this feature, separated into clearer commits. I can do that tomorrow.
